src/Microsoft.VisualStudio.Threading.Tests/AsyncLazyInitializerTests.cs

```diff
@@ -0,0 +1,271 @@
+            int invocationCount = 0;
+            var lazy = new AsyncLazyInitializer(async delegate
+            {
+                invocationCount++;
+                await Task.Yield();
+                throw new InvalidOperationException();
+            });
+
+            Assert.Equal(0, invocationCount);
+            Assert.False(lazy.IsCompleted);
+            Assert.False(lazy.IsCompletedSuccessfully);
+
+            Assert.Throws<InvalidOperationException>(() => lazy.Initialize());
+            Assert.Equal(1, invocationCount);
+            Assert.True(lazy.IsCompleted);
+            Assert.False(lazy.IsCompletedSuccessfully);
+
+            Assert.Throws<InvalidOperationException>(() => lazy.Initialize());
+            Assert.Equal(1, invocationCount);
+            Assert.True(lazy.IsCompleted);
+            Assert.False(lazy.IsCompletedSuccessfully);
+        }
+
+        [Fact]
+        public async Task InitializeAsync_OnlyExecutesActionOnce_Failure()
+        {
+            int invocationCount = 0;
+            var lazy = new AsyncLazyInitializer(async delegate
+            {
+                invocationCount++;
+                await Task.Yield();
+                throw new InvalidOperationException();
+            });
+
+            Assert.Equal(0, invocationCount);
+            Assert.False(lazy.IsCompleted);
+            Assert.False(lazy.IsCompletedSuccessfully);
+
+            await Assert.ThrowsAsync<InvalidOperationException>(() => lazy.InitializeAsync());
+            Assert.Equal(1, invocationCount);
+            Assert.True(lazy.IsCompleted);
+            Assert.False(lazy.IsCompletedSuccessfully);
+
+            await Assert.ThrowsAsync<InvalidOperationException>(() => lazy.InitializeAsync());
+            Assert.Equal(1, invocationCount);
+            Assert.True(lazy.IsCompleted);
+            Assert.False(lazy.IsCompletedSuccessfully);
+        }
+
+        /// <summary>
+        /// Verifies that even after the action has been invoked
+        /// its dependency on the Main thread can be satisfied by
+        /// someone synchronously blocking on the Main thread that is
+        /// also interested in its completion.
+        /// </summary>
+        [Fact]
+        public void InitializeAsync_RequiresMainThreadHeldByOther()
+        {
+            var context = this.InitializeJTCAndSC();
+            var jtf = context.Factory;
+
+            var evt = new AsyncManualResetEvent();
+            var lazy = new AsyncLazyInitializer(
+                async delegate
+                {
+                    await evt; // use an event here to ensure it won't resume till the Main thread is blocked.
+                },
+                jtf);
+
+            var resultTask = lazy.InitializeAsync();
+            Assert.False(resultTask.IsCompleted);
+
+            var collection = context.CreateCollection();
+            var someRandomPump = context.CreateFactory(collection);
+            someRandomPump.Run(async delegate
```

Comment:
How do you know this will block the async lazy delegate from continuing after `evt.Set();`? This is the XUnit sync context I believe, do you know how that one behaves? Say it worked like threadpool, would it not see that the one thread is blocked (due to the JTF.Run), and post it to a different thread?

Reply:
> This is the XUnit sync context I believe

No, it's the one set by `SingleThreadedSynchronizationContext.New`, which is never the xunit one -- it's either a WPF one or our own mock one, IIRC.

> How do you know this will block the async lazy delegate from continuing after `evt.Set();`? 

Because I tested it: I removed the use of JTF from the product class and verified the test hangs. :)

Reply:
Ah right, missed that `InitializeJTCAndSC()`